File: Lab6/Stack.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Lab6
{
    class Node<T>
    {
        public Node(T data)
        {
            Data = data;
        }
        public T Data { get; set; }
        public Node<T> Next { get; set; }
    }
    class Stack<T>
    {
        Node<T> head;
        int count;
        private InvalidOperationException err411 = new InvalidOperationException("Error 411: Stack is empty");
        public bool IsEmpty() => count == 0;
        public int Count() => count;
        public void Push(T item)
        {
            Node<T> node = new Node<T>(item);
            node.Next = head;
            head = node;
            count++;
        }
        public T Pop()
        {
            if (IsEmpty())
                throw err411;
            Node<T> temp = head;
            head = head.Next;
            count--;
            return temp.Data;
        }
        public T Peek() => IsEmpty() ? throw err411 : head.Data;
        public Stack<T> Reverse()
        {
            Stack<T> copy = new Stack<T>();
            Node<T> node = head;
            for (; node != null; node = node.Next)
                copy.Push(node.Data);
            return copy;
        }
        public T[] ToArray()
        {
            T[] arr = new T[count];
            int i = 0;
            Node<T> node = head;
            for(; node != null; node = node.Next, i++)
                arr[i] = node.Data;
            return arr;
        }
    }
}

File: LabsJava/Lab8/src/Stack.java
import java.util.Arrays;

public class Stack<T> {
    Node<T> head;
    int count;
    private IllegalArgumentException err411 = new IllegalArgumentException("Error 411: Stack is empty");

    public boolean IsEmpty() {
        return count == 0;
    }

    public int Count() {
        return count;
    }

    public void Push(T item) {
        Node<T> node = new Node<T>(item);
        node.next = head;
        head = node;
        count++;
    }

    public T Pop() throws IllegalArgumentException {
        if (IsEmpty())
            throw err411;
        Node<T> temp = head;
        head = head.next;
        count--;
        return temp.value;
    }

    public Stack<T> Reverse() {
        Stack<T> copy = new Stack<T>();
        Node<T> node = head;
        for (; node != null; node = node.next)
            copy.Push(node.value);
        return copy;
    }

    public String[] ToArray() {
        @SuppressWarnings("unchecked")
        T[] arr = (T[]) new Object[count];
        int i = 0;
        Node<T> node = head;
        for (; node != null; node = node.next, i++)
            arr[i] = node.value;

        String[] stringArray = Arrays.copyOf(arr, arr.length, String[].class);
        return stringArray;
    }

    public T Peek() throws IllegalArgumentException {
        if (IsEmpty())
            throw err411;
        return head.value;
    }

}